src/main/java/org/jboss/netty/handler/codec/http/cookie/CookieEncoder.java

```diff
@@ -46,6 +53,11 @@ protected void validateCookie(String name, String value) {
             if ((pos = firstInvalidCookieValueOctet(unwrappedValue)) >= 0) {
                 throw new IllegalArgumentException("Cookie value contains an invalid char: " + value.charAt(pos));
             }
+
+            if ((pos = firstWarnCookieValueOctet(unwrappedValue)) >= 0) {
```

Comment:
We are certain the issue is with value and not "!" in cookie name?

Reply:
Caused by: java.lang.IllegalArgumentException: Cookie value contains an invalid char: !

The exception message says it's in the value.